DllImportGenerator/DllImportGenerator.UnitTests/CodeSnippets.cs

```diff
@@ -446,5 +460,285 @@ partial class Test
+        public static string CustomStructMarshallingParametersAndModifiers = BasicParametersAndModifiers("S") + @"
+[NativeMarshalling(typeof(Native))]
+struct S
+{
+    public bool b;
+}
+
+struct Native
+{
+    private int i;
+    public Native(S s)
+    {
+        i = s.b ? 1 : 0;
+    }
+
+    public S ToManaged() => new S { b = i != 0 };
+}
+";
+
+        public static string CustomStructMarshallingStackallocParametersAndModifiersNoRef = BasicParametersAndModifiersNoRef("S") + @"
+[NativeMarshalling(typeof(Native))]
+struct S
+{
+    public bool b;
+}
+
+struct Native
+{
+    private int i;
+    public Native(S s, System.Span<byte> b)
+    {
+        i = s.b ? 1 : 0;
+    }
+
+    public S ToManaged() => new S { b = i != 0 };
+
+    public const int StackBufferSize = 1;
+}
+";
+        public static string CustomStructMarshallingStackallocOnlyRefParameter = BasicParameterWithByRefModifier("ref", "S") + @"
+[NativeMarshalling(typeof(Native))]
+struct S
+{
+    public bool b;
+}
+
+struct Native
+{
+    private int i;
+    public Native(S s, System.Span<byte> b)
+    {
+        i = s.b ? 1 : 0;
+    }
+
+    public S ToManaged() => new S { b = i != 0 };
+
+    public const int StackBufferSize = 1;
+}
+";
+        public static string CustomStructMarshallingOptionalStackallocParametersAndModifiers = BasicParametersAndModifiers("S") + @"
+[NativeMarshalling(typeof(Native))]
+struct S
+{
+    public bool b;
+}
+
+struct Native
+{
+    private int i;
+    public Native(S s, System.Span<byte> b)
+    {
+        i = s.b ? 1 : 0;
+    }
+    public Native(S s)
+    {
+        i = s.b ? 1 : 0;
+    }
+
+    public S ToManaged() => new S { b = i != 0 };
+
+    public const int StackBufferSize = 1;
+}
+";
+
+        public static string CustomStructMarshallingStackallocValuePropertyParametersAndModifiersNoRef = BasicParametersAndModifiersNoRef("S") + @"
+[NativeMarshalling(typeof(Native))]
+struct S
+{
+    public bool b;
+}
+
+struct Native
+{
+    public Native(S s, System.Span<byte> b)
+    {
+        Value = s.b ? 1 : 0;
+    }
+
+    public S ToManaged() => new S { b = Value != 0 };
+
+    public int Value { get; set; }
+
+    public const int StackBufferSize = 1;
+}
+";
+        public static string CustomStructMarshallingValuePropertyParametersAndModifiers = BasicParametersAndModifiers("S") + @"
+[NativeMarshalling(typeof(Native))]
+struct S
+{
+    public bool b;
+}
+
+struct Native
+{
+    public Native(S s)
+    {
+        Value = s.b ? 1 : 0;
+    }
+
+    public S ToManaged() => new S { b = Value != 0 };
+
+    public int Value { get; set; }
+}
+";
+        public static string CustomStructMarshallingPinnableParametersAndModifiers = BasicParametersAndModifiers("S") + @"
+[NativeMarshalling(typeof(Native))]
+class S
+{
+    public int i;
+
+    public ref int GetPinnableReference() => ref i;
```

Comment:
I would expect `GetPinnableReference` to be on `Native` for this case. The marshaler type should be in full control for how the value is marshalled.

Reply:
The `GetPinnableReference` method on a struct type is only used when we're using marshalling defined by the type author (via the `[NativeMarshalling]` attribute instead of `[MarshalUsing]`). Allowing the `GetPinnableReference` on the user type enables people to write types that can be marshalled exclusively via pinning, like how `string` and blittable arrays support being marshalled by only using a pinning statement.

If we require the construction of the marshaler struct first, then we'd either stack allocate and copy when we don't need to (if the user provides a stackalloc constructor), heap allocate and copy when we don't need to, require the user to allocate on the first access of the `Value` property (basically assuming that the `Value` property is only accessed when a copy is needed), or a new constructor would need to be introduced that doesn't do any allocations.

I don't like the idea of forcing users to delay allocations until the `Value` property access since it seems non-intuitive and like a hole in the design.

I am thinking about allowing `GetPinnableReference` on the marshaler type instead of allowing a ref return `Value` property though. That will likely be a cleaner solution for that problem.

Reply:
> The GetPinnableReference method on a struct type is only used when we're using marshalling defined by the type author (via the [NativeMarshalling] attribute instead of [MarshalUsing])

Ok, I was hoping that the rules for marshalers defined by the type author will be exactly same as the rules for MarshalUsing. Is the only reason for this difference to save the author of marshaler from implementing a trivial forwarder for `GetPinnableRefence`, or is there more to it?

Reply:
The reason we allow the `GetPinnableReference` member on the managed type is it enables developers to author a type that has all of the optimizations that we have for marshalling a `string` to a `LPWSTR` with only one marshaller type. The `GetPinnableReference` method on the managed type is used for the pinning scenario (by value in), the `Span<byte>` constructor is used for the stack allocation scenario (by ref in), and the constructor that only takes the managed object is the fallback heap allocation scenario (by ref in/out).

If we didn't allow the `GetPinnableReference` method on the managed type to participate, then we'd have to determine when to prefer the non-stackalloc constructor over the stackalloc constructor (so we don't stack allocate), and the marshaller type author would have to delay allocating and marshalling their data to native until they know that the `GetPinnableReference` method will not be used, so it would have to done be in the `Value` property get accessor if we wanted to enable a single marshaller type to cover all scenarios. And in the current design, we only support associating one marshaler at type author time, any others need to be associated at type usage time via MarshalUsing.

Enabling the marshaller to use the managed `GetPinnableReference` method makes it very simple to describe the scenarios where each constructor and method on the native value type is called and what the author should put in each method body.

MarshalUsing does not enable the managed GetPinnableReference optimization because the marshaller referenced by a MarshalUsing likely marshals the data differently than a marshaller written by the type author. If a type author wants to define marshalling for the type, they can apply the `[NativeMarshalling]` attribute to the type; `[MarshalUsing]` is primarily for scenarios where the marshaler cannot be applied to the type because it lives in another assembly or already has a different default marshalling scheme.

Reply:
Ok, make sense. Thank you for the explanation.